libraries/auth/src/main/java/org/datatransferproject/auth/OAuth2ServiceExtension.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright 2018 The Data Transfer Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.datatransferproject.auth;
+
+import com.google.api.client.http.HttpTransport;
+import com.google.common.base.Preconditions;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.datatransferproject.api.launcher.ExtensionContext;
+import org.datatransferproject.spi.api.auth.AuthDataGenerator;
+import org.datatransferproject.spi.api.auth.AuthServiceProviderRegistry.AuthMode;
+import org.datatransferproject.spi.api.auth.extension.AuthServiceExtension;
+import org.datatransferproject.spi.cloud.storage.AppCredentialStore;
+import org.datatransferproject.types.transfer.auth.AppCredentials;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * General implementation of an {@link AuthServiceExtension} for OAuth2.  Largely exists to provide
+ * the appropriate {@link OAuth2DataGenerator} upon request.
+ */
+public class OAuth2ServiceExtension implements AuthServiceExtension {
+
+  private static final Logger logger = LoggerFactory.getLogger(OAuth2ServiceExtension.class);
+
+  private final OAuth2Config oAuth2Config;
+
+  private volatile Map<String, OAuth2DataGenerator> exportAuthDataGenerators;
+  private volatile Map<String, OAuth2DataGenerator> importAuthDataGenerators;
+
+  private AppCredentials appCredentials;
+  private HttpTransport httpTransport;
+
+  private boolean initialized = false;
+
+  public OAuth2ServiceExtension(OAuth2Config oAuth2Config) {
+    this.oAuth2Config = oAuth2Config;
```

Comment:
We should probably validate that all the data is present in the config before continuing construction.